_episodes/05-dimensionality-reduction.md

```diff
@@ -3,15 +3,156 @@ title: "Dimensionality Reduction"
+
+from sklearn import decomposition
+from sklearn import datasets
+from sklearn import manifold
+
+digits = datasets.load_digits()
+
+# Examine the dataset
+
+print(digits.data)
+
+print(digits.target)
+
+X = digits.data
+y = digits.target
+
+~~~
+{: .language-python}
+
+### Principle Component Analysis (PCA)
+
+PCA is a technique that does rotations of data in a two dimensional
+array to decompose the array into combinations vectors that are orthogonal
+and can be ordered according to the amount of information they carry.
+
+~~~
+# PCA
+pca = decomposition.PCA(n_components=2)
+pca.fit(X)
+X_pca = pca.transform(X)
+
+fig = plt.figure(1, figsize=(4, 4))
+plt.clf()
+plt.scatter(X_pca[:, 0], X_pca[:, 1], c=y, cmap=plt.cm.nipy_spectral, 
+        edgecolor='k',label=y)
+plt.colorbar(boundaries=np.arange(11)-0.5).set_ticks(np.arange(10))
+plt.savefig("pca.svg")
+~~~
+{: .language-python}
+
+![Reduction using PCA](../fig/pca.svg)
+
+### t-distributed Stochastic Neighbor Embedding (t-SNE)
+
+~~~
+# t-SNE embedding
+tsne = manifold.TSNE(n_components=2, init='pca',
+        random_state = 0)
+X_tsne = tsne.fit_transform(X)
+fig = plt.figure(1, figsize=(4, 4))
+plt.clf()
+plt.scatter(X_tsne[:, 0], X_tsne[:, 1], c=y, cmap=plt.cm.nipy_spectral,
+        edgecolor='k',label=y)
+plt.colorbar(boundaries=np.arange(11)-0.5).set_ticks(np.arange(10))
+plt.savefig("tsne.svg")
+~~~
+{: .language-python}
+
+![Reduction using t-SNE](../fig/tsne.svg)
+
+
+
+> # Exercise: Working in three dimensions
+> The above example has considered only two dimensions since humans
+> can visualize two dimensions very well. However, there can be cases
+> where a dataset requires more than two dimensions to be appropriately
+> decomposed. Modify the above programs to use three dimensions and 
+> create appropriate plots.
+> Do three dimensions allow one to better distinguish between the digits?
+>
+> > ## Solution
+> > ~~~
+> > from mpl_toolkits.mplot3d import Axes3D
+> > # PCA
+> > pca = decomposition.PCA(n_components=3)
+> > pca.fit(X)
+> > X_pca = pca.transform(X)
+> > fig = plt.figure(1, figsize=(4, 4))
+> > plt.clf()
+> > ax = fig.add_subplot(projection='3d')
+> > ax.scatter(X_pca[:, 0], X_pca[:, 1], X_pca[:, 2], c=y,
+> >           cmap=plt.cm.nipy_spectral, s=9, lw=0)
+> > plt.savefig("pca_3d.svg")
+> > ~~~
+> > {: .language-python}
+> >
+> > ![Reduction to 3 components using pca](../fig/pca_3d.svg)
+> >
+> > ~~~
+> > # t-SNE embedding
+> > tsne = manifold.TSNE(n_components=3, init='pca',
+> >         random_state = 0)
+> > X_tsne = tsne.fit_transform(X)
+> > fig = plt.figure(1, figsize=(4, 4))
+> > plt.clf()
+> > ax = fig.add_subplot(projection='3d')
+> > ax.scatter(X_tsne[:, 0], X_tsne[:, 1], X_tsne[:, 2], c=y,
+> >           cmap=plt.cm.nipy_spectral, s=9, lw=0)
+> > plt.savefig("tsne_3d.svg")
+> > ~~~
+> > {: .language-python}
+> >
+> > ![Reduction to 3 components using tsne](../fig/tsne_3d.svg)
+> >
+> >
+> {: .solution}
+{: .challenge}
+
+> # Exercise: Parameters
```

Comment:
I took a closer look and this line must be replaced with `> ## Exercise: Parameters`, that is, the header must have two `##` and not one `#`.